During this practice, the students were asked to choose a ionic solid from the Materials Project database in order to develop the following steps.

**Important note:** remember that, in order to obtain full access to the database it is necessary to create a Materials Project account. The instructions for creating an account and start exploring the database are given in the readme file of this repository.

### 1. Download the .cif file. Visualize the crystaline sctructure using ASE GUI in order to answer the following statements. In order to realize the modifications to the structure, use ASE GUI, and to visualize, use ASE GUI or VESTA, as preferred. Replicate the unit cell as considered more convenient.

The material chosen is $\text{CaTiO}_3$ (it is possible to replicate this work what any other solid from Materials Project). In this case, the structure file will be downloaded manually from the webpage. But first, let us assume that we are just entering for the first time after successfully logging in. This what it should look like:

<img src="images/mp_init_page.png" width="700" style="margin:auto"/>

In order to begin our search for the material we are going to work with, we click the `Start Exploring Materials` button, at the left side of the page, around the section in the middle, which will bring us to this new page:


<img src="images/start_exploring_button.png" width="700" style="margin:auto"/>

As we can see, there is a search bar that will allow us to type the material we are interested in obtaining. However, it is important to mention that there are different ways to look for materials. We let the reader explore at their own pace. From now on, we are going to type in the material for this example CaTiO3 and hit enter which will give us the following results:

<img src="images/materials_search.png" width="700" style="margin:auto"/>


Notice that on the left side there's a filter section, where we can make our search more detailed to what we are specifically looking for. It is worth mentioning that we encourage the reader to explore all of these functionalities to get the most out of Materials Project, but for now, we are going to click on the first element of the `Material ID` column, which is `mp-4019`, to the right of the star sign. The new page will look like this:


<img src="images/material_overview.png" width="700" style="margin:auto"/>

This page contains a vast amount of information and properties for the material that can be explored later on, but for now we need to obtain the CIF file, which we accomplish by click the last icon inside of the square of the visual representation, the one highlighted with blue:

<img src="images/cif_download.png" width="700" style="margin:auto"/>

Notice that after clicking it, it expands into several options, from which we are going to choose the second one, `CIF`. It will open the file explorer of the computer so make sure to save it in the same directory of this notebook.

With the file downloaded, we are ready to visualize the structure using ASE (remember to have installed ASE in the environment you are working in order for the code to work properly)

In [1]:
! ase gui CaTiO3.cif

Here is an image that contains the visualizations using the multiple tools mentioned:

<img src="images/ase_vesta_MP_visualization.png" width="700" style="margin:auto"/>

Remember that by opening the file in each of these programs, you can interact with the image by rotating it to see its structure from different angles and also explore other functionalities.

We are now going to modify the structure by applying vacancies, defects and impurities

**a) Create a vacancy by removing a cation. Explain the implications this would have on both the geometry (and therefore symmetry) and the electronic structure of the solid.**

The cations in the material are $\text{Ca}^{2+}$ and $\text{Ti}^{4+}$, so the vacancy can be created by removing one of these.
For this specific example, a Calcium atom is removed, as shown highlighted in the following image:

<img src="images/cation_highlighted.png" width="300" style="margin:auto"/>


After removing the cation, a comparison is made by replicating the unit cell along each axis:

<img src="images/vs_vacancy.png" width="500" style="margin:auto"/>

By removing a cation located at one of the corners of the cell, it can be observed that the geometry is significantly affected, as the unit cell will have a different arrangement than the original. This change may alter properties such as the lattice type, although in this particular case, since the atom is located at a corner, its influence on specific lattice systems (e.g., face-centered, body-centered, primitive) might not be as significant. However, it does imply that the space group symmetry of the solid could change.

Additionally, geometric characteristics like density may vary, as the number of atoms contained within a unit cell of a given volume changes.

On the other hand, removing a cation modifies the electronic structure of the solid. The removed Ca²⁺ cation contributed electrons to the molecular orbitals that were part of the solid’s original electronic structure. These contributions influenced properties such as the specific bandgap, which determines its conductive capacity. By removing a cation, this structure is disrupted, potentially leading to changes in the band structure and, consequently, variations in the bandgap. Such changes could alter the solid's conductive properties.

**b) Create a Frenkel defect. Explain.**

In a Frenkel defect, a cation from the same material is relocated to an interstitial position (between two layers). To facilitate this process, the unit cell is first replicated along two axes. Then, a cation is selected from the solid and moved to an empty space. Following the organizational pattern of the atoms, the cation is relocated to a position where it is surrounded by oxygen anions, as is typical for cations in the material. The formation of this defect is illustrated below:


<img src="images/frenkel_defect.png" width="500" style="margin:auto"/>


The implications of this defect on the solid differ slightly from those of vacancies. This is because, although an atom is removed from its original position, it is reintroduced elsewhere within the solid. This means the overall charge of the system remains unaffected, unlike in the case of a vacancy. However, it can still be considered that this rearrangement of charges may influence the behavior of the system.

Specifically, upon further investigation into the implications of this defect, it was found that it can increase the dielectric constant of the solid. This occurs because when the atom is relocated, if it ends up in a region where there are other charges similar to its own, these charges will repel each other.

**c) Create a Schottky defect. Explain.**

In a Schottky defect, vacancies occur in pairs of atoms such that when they are removed, the charge of the solid remains neutral. This means the removed pair must consist of cations and anions whose charges sum to zero. For this example, an oxygen atom with a charge of 2⁻ and a calcium atom with a charge of 2⁺ were removed. As a result, the total charge of the solid is unaffected by these vacancies.

It is also noteworthy that the removal of these pairs does not require the atoms to be adjacent; they can be taken from arbitrary parts of the cell. The process is illustrated below:

<img src="images/schottky_defect.png" width="600" style="margin:auto"/>


Similarly to simple vacancies, the removal of atoms results in a lower concentration of atoms within the cell, leading to a decrease in its density.

**d) Add an interstitial impurity from a foreign atom. Explain.**

For this type of impurity, it is necessary to introduce an atom that occupies an interstitial position, similar to how the calcium cation was positioned in part (b). In this case, the number of atoms in the defect-free solid remains unchanged, and since an additional atom is being introduced into the unit cell of a fixed volume, properties like density will change, specifically increasing.

Additionally, the electronic structure of the solid may also be affected, as the charge of the added atom will influence the charges of the surrounding atoms. This could lead to changes in the band structure and, consequently, variations in the solid's conductivity.

The selected element is Chromium (Cr), which has the following oxidation states, as noted in [Ptable](https://ptable.com/?lang=en#Properties):

- -2, -1, 1, 2, 3, 4, 5, 6

This means there are multiple possible oxidation states that the Cr atom could assume when placed in the cell. The designated position for the Cr atom is primarily surrounded by oxygen atoms, with calcium atoms nearby, and titanium atoms located in the adjacent layer. The process is illustrated below:

<img src="images/interstitial_impurity.png" width="600" style="margin:auto"/>

**e) Demonstrate an isomorphic substitution impurity by replacing one atom with a different one. Explain.**

For this type of impurity, it is essential to consider various factors before making the substitution. The term "isomorphic" implies that the atom used to replace one in the system must have similar characteristics, particularly in terms of atomic size and charge. Therefore, an exploration of the periodic table was conducted to identify elements with properties resembling those of $\text{Ca}^{2+}$.

It was observed that elements adjacent to calcium on the periodic table have significantly smaller atomic sizes compared to calcium. Consequently, elements within the transition metals that can also exhibit a $+2$ oxidation state were considered. As shown in the following image, rhodium ($Rh$) appears to be a suitable choice for substitution:

<img src="images/isomorphic_substitution.png" width="500" style="margin:auto"/>

Thus, a $\text{Ca}^{2+}$ atom was replaced with $\text{Rh}^{2+}$ as follows:

<img src="images/isomorphic_substitution_2.png" width="600" style="margin:auto"/>

### Use the structure outlined in numeral 1) to develop the following statements. In the relevant sections, include the lines of code used. 

**a) Locate the phase diagram of the crystalline system and analyze it in detail. Indicate which species is (or are) the most stable and whether the system you have is stable.**

In order to obtain the phase diagram we are going back to the Materials Project webpage where we left it, with all the information about the [material we are working with](https://next-gen.materialsproject.org/materials/mp-4019?formula=CaTiO3).

Recall that when we downloaded the CIF file, there was a menu at the left side of the page called `table of contents`. We encourage the reader to explore the different information that can be found in each of these section but for now we are going to click on the one called `Properties`, which will show us the following:

<img src="images/phase_diagram_step1.png" width="700" style="margin:auto"/>

And we are going to click button at the lowest part on the image where it says `Generate Phase Diagram`:

<img src="images/phase_diagram2.png" width="700" style="margin:auto"/>

In the lower-left corner is $\text{Ti}$, in the lower-right corner is $\text{O}_2$, and at the top is $\text{Ca}$.

The proportion of each element composing the solid being analyzed ($\text{CaTiO}_{3}$) can also be obtained from the phase diagram: 

$$\text{Ca}_{0.2}\text{Ti}_{0.2}\text{O}_{0.6}$$

The most stable species are highlighted in green. It is evident that the majority are located at the bottom of the diagram, between titanium and oxygen, suggesting that the most stable species are composed of these two elements in varying proportions (sometimes with titanium in greater proportions, and sometimes with more oxygen).

It is also noticeable that there do not appear to be species formed solely from titanium and calcium, regardless of their proportions.

Finally, the species $\text{CaTiO}_{3}$ is identified as stable.

**Locate the formation energy graph. Analyze it.**

The formation energy is related to the phase diagram presented in the previous section, as it is possible to identify the formation energies for each compound in the diagram. A pattern can be observed indicating that the compounds with lower formation energies correspond to stable compounds (those marked in green), while compounds that are not identified as stable have higher formation energies associated with them.

Upon closely examining the values of these energies, it is observed that for each compound labeled as stable, the formation energy tends to be around -3 eV, while the compounds marked with red diamonds typically have formation energies close to or greater than -2 eV.

**c) Locate the electronic structure, particularly the band structure and the density of states. Analyze the graph according to what was covered in class.**

In order to obtain the band structure and the density of states, we are going to click on `Electronic Structure`, next to `Phase Stability`, where we should be located after generating the phase diagram. 

<img src="images/electronic_structure_step1.png" width="700" style="margin:auto"/>

We will show the band structure more clearly in order to analyze it properly:

<img src="images/electronic_structure.png" width="700" style="margin:auto"/>

Upon closely examining the band structure, we can identify the valence band at the bottom and the conduction band at the top. A gap is observed between them, which initially suggests that the solid is not a metal (conductor). Therefore, it is necessary to consider the size of this gap in more detail to determine whether it is a semiconductor or an insulator. According to the information provided by the Materials Project, the bandgap of this solid is 2.31 eV, which, based on the intervals discussed in class, indicates that the solid is a semiconductor.

Additionally, the density of states reveals the elements associated with the conduction and valence bands. For this solid, oxygen predominantly occupies the valence band, while calcium and titanium predominantly occupy the conduction band. This observation highlights a trend where cations tend to move toward the conduction band, while anions are more likely to be found in the valence band. This can be interpreted as the valence band containing orbitals that have a complete number of electrons, while the conduction band corresponds to orbitals that are not fully occupied.



**d) Using Jupyter Lab, load the .CIF file, visualize (print) the crystallographic information, and observe the structure using the Crystal Toolkit. Determine the oxidation states of each atom in the crystalline structure using the methodology discussed in class. Visualize your data (print). Show the lines of code used.**

Even though we already have the CIF file of our material because we manually downloaded it from Materials Project, we are going to show the step to use the Materials Project API to easily obtain the structure with just a few lines of code.

First of all, please make sure you have all the necessary libraries installed (see the readme file of this repository for guidance)

In [1]:
#import libraries
import pymatgen
import crystal_toolkit
from pymatgen.ext.matproj import MPRester
from pymatgen.transformations.standard_transformations import AutoOxiStateDecorationTransformation #to determine the oxidation states

The next step is to obtain the API key. This is necessary to make any request to the Materials Project API and there's a unique key for each Materials Project account. **You shouldn't share this key with anyone.**

In order to obtain the API key, we are going to the main page of Materials Project after logging in, which should look like this:

<img src="images/mp_init_page.png" width="700" style="margin:auto"/>

Then we go to the top right and click on the `API` button and it will shows us this new page:


<img src="images/api_page.png" width="700" style="margin:auto"/>

Note the menu on the left side on the screen. Click on the `API Key` button at it will show you something like this:

<img src="images/api_key.png" width="700" style="margin:auto"/>

Your personal API key should appear where the red line is located in the image. You can select and copy it, and then paste it in the following cell as indicated:

In [2]:
# In this cell you are going to paste your API key
"""
The next two lines of code should be uncommented and executed if you want to try
the steps by yourself locally using the API and your personal key.
In order to continue with this example, we only read the structure from the CIF
file previously downloaded, however, after readind the structure either from the
CIF file or the API, the rest of the process will be the same
"""
# mpr = MPRester('your API key should go here') #replace the text inside the quotation marks with your key 
# CaTiO3 = mpr.get_structure_by_material_id('mp-4019')

"""
So if you are going to use your API key, the next two lines should be commented
because they just read the structure from the CIF file we already
downloaded manually
"""
from pymatgen.core.structure import Structure

# Load the CIF file
CaTiO3 = Structure.from_file("CaTiO3.cif")

Now that we have our structure, we want to see the crystallographic information as follows:

In [3]:
#crystallographic information
print(CaTiO3)

Full Formula (Ca4 Ti4 O12)
Reduced Formula: CaTiO3
abc   :   5.372049   5.462796   7.636360
angles:  90.000000  90.000000  90.000000
pbc   :       True       True       True
Sites (20)
  #  SP           a         b         c
---  ----  --------  --------  --------
  0  Ca2+  0.509067  0.542398  0.75
  1  Ca2+  0.490933  0.457602  0.25
  2  Ca2+  0.990933  0.042398  0.75
  3  Ca2+  0.009067  0.957602  0.25
  4  Ti4+  0         0.5       0.5
  5  Ti4+  0         0.5       0
  6  Ti4+  0.5       0         0.5
  7  Ti4+  0.5       0         0
  8  O2-   0.076781  0.481328  0.75
  9  O2-   0.923219  0.518672  0.25
 10  O2-   0.423219  0.981328  0.75
 11  O2-   0.576781  0.018672  0.25
 12  O2-   0.790294  0.789614  0.959841
 13  O2-   0.209706  0.210386  0.459841
 14  O2-   0.209706  0.210386  0.040159
 15  O2-   0.790294  0.789614  0.540159
 16  O2-   0.709706  0.289614  0.959841
 17  O2-   0.290294  0.710386  0.459841
 18  O2-   0.290294  0.710386  0.040159
 19  O2-   0.709706  0.289614  

Then we visualize the structure using Crystal Toolkit

In [4]:
CaTiO3

If you see this text, the Crystal Toolkit Jupyter Lab 

extension is not installed. You can install it by running 

"pip install crystaltoolkit-extension" 

from the same environment you run "jupyter lab". 

This only works in Jupyter Lab 3.x or above.


Structure Summary
Lattice
    abc : 5.37204886 5.46279649 7.63635962
 angles : 90.0 90.0 90.0
 volume : 224.09973769300066
      A : 5.37204886 0.0 3.2894312206310936e-16
      B : -3.344998117935948e-16 5.46279649 3.344998117935948e-16
      C : 0.0 0.0 7.63635962
    pbc : True True True
PeriodicSite: Ca0 (Ca2+) (2.735, 2.963, 5.727) [0.5091, 0.5424, 0.75]
PeriodicSite: Ca1 (Ca2+) (2.637, 2.5, 1.909) [0.4909, 0.4576, 0.25]
PeriodicSite: Ca2 (Ca2+) (5.323, 0.2316, 5.727) [0.9909, 0.0424, 0.75]
PeriodicSite: Ca3 (Ca2+) (0.04871, 5.231, 1.909) [0.009067, 0.9576, 0.25]
PeriodicSite: Ti4 (Ti4+) (-1.672e-16, 2.731, 3.818) [0.0, 0.5, 0.5]
PeriodicSite: Ti5 (Ti4+) (-1.672e-16, 2.731, 1.672e-16) [0.0, 0.5, 0.0]
PeriodicSite: Ti6 (Ti4+) (2.686

The image is interactive so you can rotate it, zoom in and out

Then we want to determine the oxidation states of each atom in the crystalline structure. For this we use the module we previously imported `AutoOxiStateDecorationTransformation`

In [5]:
#Oxidation states of each atom
chg= AutoOxiStateDecorationTransformation()
print(chg.apply_transformation(CaTiO3))

Full Formula (Ca4 Ti4 O12)
Reduced Formula: CaTiO3
abc   :   5.372049   5.462796   7.636360
angles:  90.000000  90.000000  90.000000
pbc   :       True       True       True
Sites (20)
  #  SP           a         b         c
---  ----  --------  --------  --------
  0  Ca2+  0.509067  0.542398  0.75
  1  Ca2+  0.490933  0.457602  0.25
  2  Ca2+  0.990933  0.042398  0.75
  3  Ca2+  0.009067  0.957602  0.25
  4  Ti4+  0         0.5       0.5
  5  Ti4+  0         0.5       0
  6  Ti4+  0.5       0         0.5
  7  Ti4+  0.5       0         0
  8  O2-   0.076781  0.481328  0.75
  9  O2-   0.923219  0.518672  0.25
 10  O2-   0.423219  0.981328  0.75
 11  O2-   0.576781  0.018672  0.25
 12  O2-   0.790294  0.789614  0.959841
 13  O2-   0.209706  0.210386  0.459841
 14  O2-   0.209706  0.210386  0.040159
 15  O2-   0.790294  0.789614  0.540159
 16  O2-   0.709706  0.289614  0.959841
 17  O2-   0.290294  0.710386  0.459841
 18  O2-   0.290294  0.710386  0.040159
 19  O2-   0.709706  0.289614  

**e) Visualize (print) the space group, volume, and symmetry. Show the lines of code used.**

In [6]:
#space group and symmetry
CaTiO3.get_space_group_info()

('Pnma', 62)

In [7]:
#volume
print(CaTiO3.volume)

224.09973769300066


In [8]:
#density
CaTiO3.density

4.02925941391082

**f) Select an atom of your choice and move it 0.1 Å in the direction of your preference. Indicate the type of spatial symmetry this system would have, as well as the new density and volume.**

We are going to create new variable with the initial same structure and then modify it.

In [15]:
# use the following line if you are working with the API
# CaTiO3_change_location = mpr.get_structure_by_material_id('mp-4019')


# comment the following line if you are working with the API
CaTiO3_change_location = Structure.from_file("CaTiO3.cif")


In order the move an atom we use the method `translate_sites`, its arguments are the atom that is going to be moved, its direction and the distance.

We indicate the atom with the index obtained by printing the crystallographic information of the structure as we previously did:

In [16]:
print(CaTiO3_change_location)

Full Formula (Ca4 Ti4 O12)
Reduced Formula: CaTiO3
abc   :   5.372049   5.462796   7.636360
angles:  90.000000  90.000000  90.000000
pbc   :       True       True       True
Sites (20)
  #  SP           a         b         c
---  ----  --------  --------  --------
  0  Ca2+  0.509067  0.542398  0.75
  1  Ca2+  0.490933  0.457602  0.25
  2  Ca2+  0.990933  0.042398  0.75
  3  Ca2+  0.009067  0.957602  0.25
  4  Ti4+  0         0.5       0.5
  5  Ti4+  0         0.5       0
  6  Ti4+  0.5       0         0.5
  7  Ti4+  0.5       0         0
  8  O2-   0.076781  0.481328  0.75
  9  O2-   0.923219  0.518672  0.25
 10  O2-   0.423219  0.981328  0.75
 11  O2-   0.576781  0.018672  0.25
 12  O2-   0.790294  0.789614  0.959841
 13  O2-   0.209706  0.210386  0.459841
 14  O2-   0.209706  0.210386  0.040159
 15  O2-   0.790294  0.789614  0.540159
 16  O2-   0.709706  0.289614  0.959841
 17  O2-   0.290294  0.710386  0.459841
 18  O2-   0.290294  0.710386  0.040159
 19  O2-   0.709706  0.289614  

From which we are going to move atom `3`, that corresponds to Calcium. Then for the distance and direction, it will be determined as a list of three elements and each element will be the distance along direction x, y and z.

In this case, as we are going to move the Calcium atom 0.1 Å along the z axis, the argument should as this:

In [18]:
CaTiO3_change_location.translate_sites([3],[0,0,0.1])

If you see this text, the Crystal Toolkit Jupyter Lab 

extension is not installed. You can install it by running 

"pip install crystaltoolkit-extension" 

from the same environment you run "jupyter lab". 

This only works in Jupyter Lab 3.x or above.


Structure Summary
Lattice
    abc : 5.37204886 5.46279649 7.63635962
 angles : 90.0 90.0 90.0
 volume : 224.09973769300066
      A : 5.37204886 0.0 3.2894312206310936e-16
      B : -3.344998117935948e-16 5.46279649 3.344998117935948e-16
      C : 0.0 0.0 7.63635962
    pbc : True True True
PeriodicSite: Ca0 (Ca2+) (2.735, 2.963, 5.727) [0.5091, 0.5424, 0.75]
PeriodicSite: Ca1 (Ca2+) (2.637, 2.5, 1.909) [0.4909, 0.4576, 0.25]
PeriodicSite: Ca2 (Ca2+) (5.323, 0.2316, 5.727) [0.9909, 0.0424, 0.75]
PeriodicSite: Ca3 (Ca2+) (0.04871, 5.231, 2.673) [0.009067, 0.9576, 0.35]
PeriodicSite: Ti4 (Ti4+) (-1.672e-16, 2.731, 3.818) [0.0, 0.5, 0.5]
PeriodicSite: Ti5 (Ti4+) (-1.672e-16, 2.731, 1.672e-16) [0.0, 0.5, 0.0]
PeriodicSite: Ti6 (Ti4+) (2.686

In [19]:
CaTiO3_change_location.get_space_group_info()

('P1', 1)

It is possible to see that the spatial symmetry is quite different from the original, as it changed from an orthorhombic Bravais lattice to a triclinic one.

In [20]:
CaTiO3_change_location.volume

224.09973769300066

In [21]:
CaTiO3_change_location.density

4.02925941391082

On the other hand, the volume and density remain constant.

**g) Select a different atom and replace it with another type of atom of your choice. Indicate the resulting type of symmetry, along with the new density and volume.**

To replace one atom with another, the same selection process is carried out to choose an atom with the same oxidation state and a similar atomic size. In this case, one of the titanium atoms is replaced with cobalt, as shown in the following image, since they have a very similar atomic size, and cobalt has a 4+ oxidation state among its possible states:

<img src="images/substitution.png" width="400" style="margin:auto"/>



Again, we create another variable with the same original structure to modify it

In [22]:
# use the following line if you are working with the API
# CaTiO3_change_location = mpr.get_structure_by_material_id('mp-4019')


# comment the following line if you are working with the API
CaTiO3_substitution = Structure.from_file("CaTiO3.cif")

As we are going to replace one titanium atom, we are going to check the indices of these types of atoms

In [24]:
print(CaTiO3_substitution)

Full Formula (Ca4 Ti4 O12)
Reduced Formula: CaTiO3
abc   :   5.372049   5.462796   7.636360
angles:  90.000000  90.000000  90.000000
pbc   :       True       True       True
Sites (20)
  #  SP           a         b         c
---  ----  --------  --------  --------
  0  Ca2+  0.509067  0.542398  0.75
  1  Ca2+  0.490933  0.457602  0.25
  2  Ca2+  0.990933  0.042398  0.75
  3  Ca2+  0.009067  0.957602  0.25
  4  Ti4+  0         0.5       0.5
  5  Ti4+  0         0.5       0
  6  Ti4+  0.5       0         0.5
  7  Ti4+  0.5       0         0
  8  O2-   0.076781  0.481328  0.75
  9  O2-   0.923219  0.518672  0.25
 10  O2-   0.423219  0.981328  0.75
 11  O2-   0.576781  0.018672  0.25
 12  O2-   0.790294  0.789614  0.959841
 13  O2-   0.209706  0.210386  0.459841
 14  O2-   0.209706  0.210386  0.040159
 15  O2-   0.790294  0.789614  0.540159
 16  O2-   0.709706  0.289614  0.959841
 17  O2-   0.290294  0.710386  0.459841
 18  O2-   0.290294  0.710386  0.040159
 19  O2-   0.709706  0.289614  

We see that their indices are from 4 to 7, so we are going to choose 6 as an example. We are going to use the method `replace` and it receives as parameters the index of the atom we want to replace and the species of the atom that will take that replacement, as follows:

In [25]:
# replace the atom of index 6(titanium) with Cobalt
CaTiO3_substitution.replace(6, 'Co')

If you see this text, the Crystal Toolkit Jupyter Lab 

extension is not installed. You can install it by running 

"pip install crystaltoolkit-extension" 

from the same environment you run "jupyter lab". 

This only works in Jupyter Lab 3.x or above.


Structure Summary
Lattice
    abc : 5.37204886 5.46279649 7.63635962
 angles : 90.0 90.0 90.0
 volume : 224.09973769300066
      A : 5.37204886 0.0 3.2894312206310936e-16
      B : -3.344998117935948e-16 5.46279649 3.344998117935948e-16
      C : 0.0 0.0 7.63635962
    pbc : True True True
PeriodicSite: Ca0 (Ca2+) (2.735, 2.963, 5.727) [0.5091, 0.5424, 0.75]
PeriodicSite: Ca1 (Ca2+) (2.637, 2.5, 1.909) [0.4909, 0.4576, 0.25]
PeriodicSite: Ca2 (Ca2+) (5.323, 0.2316, 5.727) [0.9909, 0.0424, 0.75]
PeriodicSite: Ca3 (Ca2+) (0.04871, 5.231, 1.909) [0.009067, 0.9576, 0.25]
PeriodicSite: Ti4 (Ti4+) (-1.672e-16, 2.731, 3.818) [0.0, 0.5, 0.5]
PeriodicSite: Ti5 (Ti4+) (-1.672e-16, 2.731, 1.672e-16) [0.0, 0.5, 0.0]
PeriodicSite: Co (2.686, 0.0, 3

In [26]:
CaTiO3_substitution.get_space_group_info()

('P-1', 2)

In [27]:
print(CaTiO3_substitution)

Full Formula (Ca4 Ti3 Co1 O12)
Reduced Formula: Ca4Ti3CoO12
abc   :   5.372049   5.462796   7.636360
angles:  90.000000  90.000000  90.000000
pbc   :       True       True       True
Sites (20)
  #  SP           a         b         c
---  ----  --------  --------  --------
  0  Ca2+  0.509067  0.542398  0.75
  1  Ca2+  0.490933  0.457602  0.25
  2  Ca2+  0.990933  0.042398  0.75
  3  Ca2+  0.009067  0.957602  0.25
  4  Ti4+  0         0.5       0.5
  5  Ti4+  0         0.5       0
  6  Co    0.5       0         0.5
  7  Ti4+  0.5       0         0
  8  O2-   0.076781  0.481328  0.75
  9  O2-   0.923219  0.518672  0.25
 10  O2-   0.423219  0.981328  0.75
 11  O2-   0.576781  0.018672  0.25
 12  O2-   0.790294  0.789614  0.959841
 13  O2-   0.209706  0.210386  0.459841
 14  O2-   0.209706  0.210386  0.040159
 15  O2-   0.790294  0.789614  0.540159
 16  O2-   0.709706  0.289614  0.959841
 17  O2-   0.290294  0.710386  0.459841
 18  O2-   0.290294  0.710386  0.040159
 19  O2-   0.709706  0

Note that the sixth atom is now Co

In [28]:
CaTiO3_substitution.volume

224.09973769300066

In [29]:
CaTiO3_substitution.density

4.111257944142422

Density increased but the volume remains constant

**h) Remove an arbitrary atom and show the final symmetry, as well as the new density and volume**

In [30]:
# use the following line if you are working with the API
# CaTiO3_change_location = mpr.get_structure_by_material_id('mp-4019')


# comment the following line if you are working with the API
CaTiO3_vacancy = Structure.from_file("CaTiO3.cif")

In order to remove an atom, we use the method `remove_sites`, which receives a list with the indices of the atoms that are going to be removed. We take a look and choose the atom:

In [31]:
print(CaTiO3_vacancy)

Full Formula (Ca4 Ti4 O12)
Reduced Formula: CaTiO3
abc   :   5.372049   5.462796   7.636360
angles:  90.000000  90.000000  90.000000
pbc   :       True       True       True
Sites (20)
  #  SP           a         b         c
---  ----  --------  --------  --------
  0  Ca2+  0.509067  0.542398  0.75
  1  Ca2+  0.490933  0.457602  0.25
  2  Ca2+  0.990933  0.042398  0.75
  3  Ca2+  0.009067  0.957602  0.25
  4  Ti4+  0         0.5       0.5
  5  Ti4+  0         0.5       0
  6  Ti4+  0.5       0         0.5
  7  Ti4+  0.5       0         0
  8  O2-   0.076781  0.481328  0.75
  9  O2-   0.923219  0.518672  0.25
 10  O2-   0.423219  0.981328  0.75
 11  O2-   0.576781  0.018672  0.25
 12  O2-   0.790294  0.789614  0.959841
 13  O2-   0.209706  0.210386  0.459841
 14  O2-   0.209706  0.210386  0.040159
 15  O2-   0.790294  0.789614  0.540159
 16  O2-   0.709706  0.289614  0.959841
 17  O2-   0.290294  0.710386  0.459841
 18  O2-   0.290294  0.710386  0.040159
 19  O2-   0.709706  0.289614  

As an example, we are going to remove one of the 4 calcium atoms, the one with index 1:

In [32]:
CaTiO3_vacancy.remove_sites([1])

If you see this text, the Crystal Toolkit Jupyter Lab 

extension is not installed. You can install it by running 

"pip install crystaltoolkit-extension" 

from the same environment you run "jupyter lab". 

This only works in Jupyter Lab 3.x or above.


Structure Summary
Lattice
    abc : 5.37204886 5.46279649 7.63635962
 angles : 90.0 90.0 90.0
 volume : 224.09973769300066
      A : 5.37204886 0.0 3.2894312206310936e-16
      B : -3.344998117935948e-16 5.46279649 3.344998117935948e-16
      C : 0.0 0.0 7.63635962
    pbc : True True True
PeriodicSite: Ca0 (Ca2+) (2.735, 2.963, 5.727) [0.5091, 0.5424, 0.75]
PeriodicSite: Ca2 (Ca2+) (5.323, 0.2316, 5.727) [0.9909, 0.0424, 0.75]
PeriodicSite: Ca3 (Ca2+) (0.04871, 5.231, 1.909) [0.009067, 0.9576, 0.25]
PeriodicSite: Ti4 (Ti4+) (-1.672e-16, 2.731, 3.818) [0.0, 0.5, 0.5]
PeriodicSite: Ti5 (Ti4+) (-1.672e-16, 2.731, 1.672e-16) [0.0, 0.5, 0.0]
PeriodicSite: Ti6 (Ti4+) (2.686, 0.0, 3.818) [0.5, 0.0, 0.5]
PeriodicSite: Ti7 (Ti4+) (2.686, 0.0, 

In [33]:
print(CaTiO3_vacancy)

Full Formula (Ca3 Ti4 O12)
Reduced Formula: Ca3Ti4O12
abc   :   5.372049   5.462796   7.636360
angles:  90.000000  90.000000  90.000000
pbc   :       True       True       True
Sites (19)
  #  SP           a         b         c
---  ----  --------  --------  --------
  0  Ca2+  0.509067  0.542398  0.75
  1  Ca2+  0.990933  0.042398  0.75
  2  Ca2+  0.009067  0.957602  0.25
  3  Ti4+  0         0.5       0.5
  4  Ti4+  0         0.5       0
  5  Ti4+  0.5       0         0.5
  6  Ti4+  0.5       0         0
  7  O2-   0.076781  0.481328  0.75
  8  O2-   0.923219  0.518672  0.25
  9  O2-   0.423219  0.981328  0.75
 10  O2-   0.576781  0.018672  0.25
 11  O2-   0.790294  0.789614  0.959841
 12  O2-   0.209706  0.210386  0.459841
 13  O2-   0.209706  0.210386  0.040159
 14  O2-   0.790294  0.789614  0.540159
 15  O2-   0.709706  0.289614  0.959841
 16  O2-   0.290294  0.710386  0.459841
 17  O2-   0.290294  0.710386  0.040159
 18  O2-   0.709706  0.289614  0.540159


See how now we are left with 3 calcium atoms

In [34]:
CaTiO3_vacancy.get_space_group_info()

('Pm', 6)

In [35]:
CaTiO3_vacancy.volume

224.09973769300066

In [36]:
CaTiO3_vacancy.density

3.7322885856702057

Similarly to the substitution, the volume remains constant, but the density changes; in this case, it decreases, which makes sense since having one less atom means the cell contains less mass.

Here is a table that summarizes all the changes for each modified structure:


<img src="images/summary_defects.png" width="600" style="margin:auto"/>

According to the results obtained for the space group of each solid with defects, a significant difference can be observed compared to the original structure, as all the defects caused the system to change from orthorhombic to either triclinic or monoclinic. On the other hand, the volume does not change for any of the defects, which makes sense since the cell in which the atoms are distributed remains the same. Based on this constant volume factor, another value that did vary can be identified: the density.

First, after the translation, the density remains the same as that of the original solid, which makes sense since no atoms were removed or replaced with others of a different species; the position of an atom was simply changed within the same cell. Then, with the substitution, it is observed that the density increased slightly, which can be explained by the substituted atom. Although an element with similar characteristics in terms of size and charge was chosen, there are still differences compared to the original atom, which accounts for the increase in density. Finally, as expected, when the number of atoms within the cell decreases due to a vacancy, the density decreases.

### Use your crystal structure from the previous exercise to employ it in Pymatgen for property search criteria. In the relevant sections, show the lines of code used.

**a) Locate in the Periodic Table of Elements the metals neighboring the transition metal in your structure and choose one of them. Use as a search criterion a band gap between 1.1 and 3.2 eV or a band gap between 0.4 and 1.0 eV (use only one range) to determine the number of materials that meet this criterion. Additionally, filter the search by type of structure (chemsys). Look for materials with compositional similarity to your initial material. For example, if you start with an iron oxide (Fe3O4), search for oxides of Mn, Ru, or Co, or combinations of these if they exist—for example, a double oxide of Mn and Co. Create a DataFrame with the obtained results.**

Note that we will be using `Pymatgen` to connect with the Materials Project API, which means that, in order to run this locally, you will have to use your personal API key as we suggest in previous exercises of this same notebook.

In [1]:
from pymatgen.ext.matproj import MPRester
import pandas as pd #to create the dataframe

In [ ]:
# please modify the following line of code before running it
mpr = MPRester('your API key should go here') #replace the text inside the quotation marks with your key 

Vanadium is a transition metal adjacent to Titanium.
The band gap range will be 1.1 to 3.2 eV since this range corresponds to the band gaps of materials used for solar cells.

In [ ]:
data_v_ti = mpr.summary.search(chemsys=["Ca-Ti-V-O"], band_gap=[1.1, 3.2])

In [ ]:
df_v_ti=pd.DataFrame(data_v_ti)

In [ ]:
df_v_ti

In [ ]:
data_v = mpr.summary.search(chemsys=["Ca-V-O"], band_gap=[1.1, 3.2])

In [ ]:
df_v = pd.DataFrame(data_v)

In [ ]:
df_v

Zirconium is also a transition metal close to Titanium.

In [ ]:
data_zr = mpr.summary.search(chemsys=["Ca-Zr-O"], band_gap=[1.1, 3.2])

In [ ]:
df_zr = pd.DataFrame(data_zr)

In [ ]:
df_zr

In [ ]:
data_sc = mpr.summary.search(chemsys=["Ca-Sc-O"], band_gap=[1.1, 3.2])

A search was also conducted with Scandium, but no results were obtained.

In [ ]:
#unify the df obtained with Pymatgen
pymatgen_df = pd.concat([df_v_ti, df_v,df_zr])

In [ ]:
pymatgen_df

Now we will use Matminer, which doesn't require an API key. Just import the library that was previously downloaded during the first steps for setting up the environment (see readme file)

In [2]:
import matminer
from matminer.datasets import get_available_datasets
from matminer.datasets import load_dataset

The module `get_available_datasets` will show us all the datasets of the library with a brief description of each of them in order to choose an appropiate one for our search

In [3]:
get_available_datasets()

boltztrap_mp: Effective mass and thermoelectric properties of 8924 compounds in The  Materials Project database that are calculated by the BoltzTraP software package run on the GGA-PBE or GGA+U density functional theory calculation results. The properties are reported at the temperature of 300 Kelvin and the carrier concentration of 1e18 1/cm3.

brgoch_superhard_training: 2574 materials used for training regressors that predict shear and bulk modulus.

castelli_perovskites: 18,928 perovskites generated with ABX combinatorics, calculating gllbsc band gap and pbe structure, and also reporting absolute band edge positions and heat of formation.

citrine_thermal_conductivity: Thermal conductivity of 872 compounds measured experimentally and retrieved from Citrine database from various references. The reported values are measured at various temperatures of which 295 are at room temperature.

dielectric_constant: 1,056 structures with dielectric properties, calculated with DFPT-PBE.

double_

['boltztrap_mp',
 'brgoch_superhard_training',
 'castelli_perovskites',
 'citrine_thermal_conductivity',
 'dielectric_constant',
 'double_perovskites_gap',
 'double_perovskites_gap_lumo',
 'elastic_tensor_2015',
 'expt_formation_enthalpy',
 'expt_formation_enthalpy_kingsbury',
 'expt_gap',
 'expt_gap_kingsbury',
 'flla',
 'glass_binary',
 'glass_binary_v2',
 'glass_ternary_hipt',
 'glass_ternary_landolt',
 'heusler_magnetic',
 'jarvis_dft_2d',
 'jarvis_dft_3d',
 'jarvis_ml_dft_training',
 'm2ax',
 'matbench_dielectric',
 'matbench_expt_gap',
 'matbench_expt_is_metal',
 'matbench_glass',
 'matbench_jdft2d',
 'matbench_log_gvrh',
 'matbench_log_kvrh',
 'matbench_mp_e_form',
 'matbench_mp_gap',
 'matbench_mp_is_metal',
 'matbench_perovskites',
 'matbench_phonons',
 'matbench_steels',
 'mp_all_20181018',
 'mp_nostruct_20181018',
 'phonon_dielectric_mp',
 'piezoelectric_tensor',
 'ricci_boltztrap_mp_tabular',
 'steel_strength',
 'superconductivity2018',
 'tholander_nitrides',
 'ucsb_thermoe

We see that there are several to choose from, so we are first going to take a look at the `castelli_perovskites` dataset as we are working with a perovskite material too

In [4]:
df_matminer = load_dataset('castelli_perovskites',data_home='.')

Then we check what are the columns (the properties) this dataset contains, as we need to focus on the bandgap

In [5]:
df_matminer.columns

Index(['fermi level', 'fermi width', 'e_form', 'gap is direct', 'structure',
       'mu_b', 'formula', 'vbm', 'cbm', 'gap gllbsc'],
      dtype='object')

In [7]:
df_matminer['gap gllbsc'].unique()

array([0.  , 0.6 , 0.5 , 0.4 , 2.  , 2.8 , 1.3 , 3.  , 1.5 , 0.3 , 1.2 ,
       4.2 , 2.2 , 0.9 , 1.9 , 6.4 , 0.2 , 5.5 , 0.7 , 3.7 , 1.  , 5.7 ,
       1.6 , 2.5 , 4.7 , 1.8 , 2.4 , 1.7 , 2.7 , 0.8 , 6.5 , 3.4 , 4.6 ,
       4.1 , 2.1 , 3.2 , 4.  , 3.1 , 1.1 , 4.3 , 3.8 , 2.6 , 1.4 , 2.9 ,
       6.2 , 2.3 , 6.3 , 5.2 , 5.8 , 3.5 , 5.9 , 6.9 , 4.4 , 4.5 , 6.6 ,
       5.4 , 3.6 , 5.3 , 5.1 , 5.6 , 3.9 , 3.3 , 3.05, 6.7 , 6.  , 7.  ])

Then we check another dataset `double_perovskites_gap`

In [8]:
df_matminer = load_dataset('double_perovskites_gap',data_home='.')

In [9]:
df_matminer.columns

Index(['formula', 'a_1', 'b_1', 'a_2', 'b_2', 'gap gllbsc'], dtype='object')

In [10]:
df_matminer['gap gllbsc']

0       4.164543
1       4.454629
2       4.192522
3       3.963857
4       2.881239
          ...   
1301    2.425122
1302    2.297476
1303    3.507231
1304    1.877169
1305    2.819711
Name: gap gllbsc, Length: 1306, dtype: float64

Then we filter our search by the bandgap limits

In [11]:
bandgap_bounded = df_matminer[(df_matminer['gap gllbsc']>1.1) & (df_matminer['gap gllbsc']<3.2)]

In [12]:
bandgap_bounded

formula a_1 b_1 a_2 b_2  gap gllbsc
4     AgNbSnTiO6  Ag  Nb  Sn  Ti    2.881239
14     AgTaCsVO6  Ag  Ta  Cs   V    2.169554
15    AgTaGaNbO6  Ag  Ta  Ga  Nb    1.784718
16    AgTaGaTaO6  Ag  Ta  Ga  Ta    2.285462
23     AgTaLiVO6  Ag  Ta  Li   V    1.661857
...          ...  ..  ..  ..  ..         ...
1300  TlTaSnTiO6  Tl  Ta  Sn  Ti    1.850790
1301  TlTaSrGeO6  Tl  Ta  Sr  Ge    2.425122
1302  TlTaSrSiO6  Tl  Ta  Sr  Si    2.297476
1304  TlTaSrTiO6  Tl  Ta  Sr  Ti    1.877169
1305   TlTaYAlO6  Tl  Ta   Y  Al    2.819711

[372 rows x 6 columns]

To facilitate the search for materials with compositions similar to CaTiO3, regular expressions are used.

In [13]:
bandgap_filtered = bandgap_bounded[bandgap_bounded['formula'].str.contains(r'^Ca.*O.*', case=True, regex=True)]

In [14]:
bandgap_filtered

formula a_1 b_1 a_2 b_2  gap gllbsc
207   CaHfCsVO6  Ca  Hf  Cs   V    2.234428
208  CaHfGaNbO6  Ca  Hf  Ga  Nb    2.259002
209  CaHfGaTaO6  Ca  Hf  Ga  Ta    3.125645
210  CaHfInNbO6  Ca  Hf  In  Nb    1.454859
211  CaHfInTaO6  Ca  Hf  In  Ta    2.030900
219   CaHfLiVO6  Ca  Hf  Li   V    2.939416
225   CaHfNaVO6  Ca  Hf  Na   V    2.576283
239  CaHfTlNbO6  Ca  Hf  Tl  Nb    2.785947
249  CaSiSnTiO6  Ca  Si  Sn  Ti    2.549361
257  CaSnSnTiO6  Ca  Sn  Sn  Ti    2.728013
264  CaTiSnTiO6  Ca  Ti  Sn  Ti    2.795445
275   CaZrCsVO6  Ca  Zr  Cs   V    2.604377
276  CaZrGaNbO6  Ca  Zr  Ga  Nb    2.228163
277  CaZrGaTaO6  Ca  Zr  Ga  Ta    3.014914
278  CaZrInNbO6  Ca  Zr  In  Nb    1.456944
279  CaZrInTaO6  Ca  Zr  In  Ta    2.057918
288   CaZrLiVO6  Ca  Zr  Li   V    3.120800
294   CaZrNaVO6  Ca  Zr  Na   V    2.457279
307  CaZrTlNbO6  Ca  Zr  Tl  Nb    2.711079

However, despite significantly narrowing down the results, some elements that are not very close to Titanium are still present, so a manual selection is made.

In [15]:
new_df = bandgap_filtered.loc[[257, 264, 275]]
new_df

formula a_1 b_1 a_2 b_2  gap gllbsc
257  CaSnSnTiO6  Ca  Sn  Sn  Ti    2.728013
264  CaTiSnTiO6  Ca  Ti  Sn  Ti    2.795445
275   CaZrCsVO6  Ca  Zr  Cs   V    2.604377

**Comparison of the DataFrames**

During the development of this task, some key characteristics of each tool became evident. However, I believe that more varied and precise results can be obtained using Pymatgen. This is because I see the use of datasets in Matminer as a limitation when searching based on specific criteria. Furthermore, I find it inconvenient to have to read through all of Matminer's datasets to determine which one might be suitable for the search. Additionally, the descriptions of those datasets are not entirely reliable concerning the recorded data, as was the case with the 'castelli_perovskites' dataset. Although it claimed to contain information on the band gap, it appears that this column did not have the correct data. On the other hand, the search process with Pymatgen was more straightforward, yielding an appropriate number of results. Compared to Matminer, Pymatgen's results adhered more rigorously to the specified filtering criteria. Meanwhile, the main issue with Matminer stemmed from the dataset used, which contained information on perovskites with multiple transition metals, whereas the focus was more on perovskites with only one metal (based on the assigned solid).

**b) Use the selected DataFrame and choose two structures. Compare these two new structures with your initial structure from point 1). Contrast their symmetries, geometry, phase diagrams, stability, and electronic configuration (band gap, density of states – band structure) within the framework of the application that you and/or your teammates presented. In other words, what behavior would these two new materials have compared to the initial material?**

The first structure we are going compare will be **Ca₂V₂O₇** and we will realize a similar search that we have previously done in Materials Project

<img src="images/Ca_search.png" width="600" style="margin:auto"/>

- symmetry:

<img src="images/symmetry_ca.png" width="600" style="margin:auto"/>

According to the information provided by the Materials Project, the solid is a triclinic system, while CaTiO3 is an orthorhombic system.

- Phase diagram

<img src="images/ca_phase_diagram.png" width="600" style="margin:auto"/>

Similar to the phase diagram of CaTiO3, most of the formed systems are located in the lower part of the diagram. Additionally, the positioning of the elements remains the same: oxygen is in the lower right corner, calcium is at the top, and the transition metal occupies the remaining position. With this in mind, it can be said that oxygen compositions with transition metals are more common, though they can also result in multiple compositions that are less stable, with high formation energies.

- Electronic configuration

<img src="images/ca_electronicp.png" width="600" style="margin:auto"/>

On one hand, it is possible to observe a band structure where the bands appear to be more organized; however, a larger gap is also visible, which, according to the Materials Project, is 2.96 eV—higher than the band gap of CaTiO3. Additionally, the density of states exhibits behavior similar to that of CaTiO3, as it shows that for the valence band, the associated element is oxygen, while for the conduction band, the elements are the two cations.

Finally, it is possible to conclude that this material, applied to solar cells, would not be a very suitable choice for study given the observed properties. This is because solar cell efficiency is only about 10% when the band gap exceeds 2.6 eV.


The second material will be $\text{CaV}_3\text{O}_7$, and we do a similar search:

<img src="images/cav3_search.png" width="600" style="margin:auto"/>

- symmetry

<img src="images/cav3_symmetry.png" width="600" style="margin:auto"/>

According to the information provided by the Materials Project, the solid has the same space group as CaTiO3, orthorhombic.

- Phase diagram

<img src="images/cav3_pd.png" width="600" style="margin:auto"/>

In this phase diagram, there is a tendency for the formed solids to move toward the lower right side, seeking a higher concentration of oxygen. Additionally, the formation energy for this solid is higher than that of the others studied, making it less stable.

- Electronic configuration

<img src="images/cav3_es.png" width="600" style="margin:auto"/>

The band structure and density of states presented for this solid are somewhat different from what has been observed so far. This is because the diagrams include colors that can provide more information about what is being observed. For the density of states, the graph shows densities for each element and also specifies spin. Moreover, the density of states does not follow the same pattern as the others, where the jumps aligned with the band structure's gap. In this case, the previously observed trend regarding the elements related to the bands is not followed either, as according to the interpretation of the graphs, the transition metal is also associated with the valence band. This behavior could provide the solid with different properties, altering its functionality when applied to solar cells. However, it can be concluded that this material is not suitable for study in potential applications with porphyrins. First, it is an unstable compound, which would cause issues with the durability and adaptability of the cells. Second, with a band gap of 2.21 eV, its efficiency is likely to be quite low.
